File: optimal/modules/optimal-module/android/src/main/java/expo/modules/optimalmodule/OptimalModule.kt
package expo.modules.optimalmodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class OptimalModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("OptimalModule")
    
    // Expose the solve function as an async function
    AsyncFunction("solve") { params: Map<String, Any> ->
      try {
        // Convert JavaScript object to Kotlin Params
        val paramsObj = convertToParams(params)
        
        // Call solve method
        val result = Optimal.solve(paramsObj)
        
        // Convert result to a Map that can be returned to JavaScript
        convertResultToMap(result)
      } catch (e: Exception) {
        mapOf(
          "error" to (e.message ?: "Unknown error")
        )
      }
    }
  }
  
  private fun convertToParams(params: Map<String, Any>): Optimal.Params {
    return Optimal.Params(
      m = (params["m"] as? Number)?.toInt() ?: 0,
      n = (params["n"] as? Number)?.toInt() ?: 0,
      k = (params["k"] as? Number)?.toInt() ?: 0,
      j = (params["j"] as? Number)?.toInt() ?: 0,
      s = (params["s"] as? Number)?.toInt() ?: 0,
      minSGroups = (params["minSGroups"] as? Number)?.toInt() ?: 1,
      seed = (params["seed"] as? Number)?.toLong(),
      toLabel = params["toLabel"] as? Boolean ?: false,
      timeoutMs = (params["timeoutMs"] as? Number)?.toLong() ?: 60000
    )
  }
  
  private fun convertResultToMap(result: Optimal.Result): Map<String, Any> {
    // Convert the samplePool to a list of serializable values
    val serializedSamplePool = result.samplePool.map { 
      when(it) {
        is String, is Number -> it
        else -> it.toString()
      }
    }
    
    // Convert the groups to a list of lists of serializable values
    val serializedGroups = result.groups.map { group ->
      group.map { item ->
        when(item) {
          is String, is Number -> item
          else -> item.toString()
        }
      }
    }
    
    return mapOf(
      "samplePool" to serializedSamplePool,
      "groups" to serializedGroups,
      "ms" to result.ms
    )
  }
}

File: optimal/modules/optimal-module/src/OptimalModule.ts
import { requireNativeModule } from 'expo-modules-core';

export interface OptimalParams {
  m: number;
  n: number;
  k: number;
  j: number;
  s: number;
  minSGroups?: number;
  seed?: number;
  toLabel?: boolean;
  timeoutMs?: number;
}

export interface OptimalResult {
  samplePool: Array<number | string>;
  groups: Array<Array<number | string>>;
  ms: number;
}


const OptimalModule = requireNativeModule('OptimalModule');

export function solve(params: OptimalParams): Promise<OptimalResult> {
  return OptimalModule.solve(
    params.m,
    params.n,
    params.k,
    params.j,
    params.s,
    params.minSGroups || 1,
    params.seed,
    params.toLabel || false,
    params.timeoutMs || 60000
  );
}
export function solveWithPool(params: OptimalParams, pool: number[]): Promise<OptimalResult> {
  return OptimalModule.solveWithPool(params, pool);
}

export default OptimalModule;
